File: prisoners/prisoners.jl
using Random
const DEBUG = false
const PRISONERS = 100
const ATTEMPTS = 50
function tryit()

    a = [1:1:PRISONERS;]
    a = shuffle(a)
    for i in 1:PRISONERS
        if DEBUG
            println("prisoner $i")
        end
        count = 0
        check = i
        while count <= ATTEMPTS
            count += 1
            if a[check] == i
                if DEBUG
                    println("Attempt $count: Checking box $check and found my number")
                end
                break
            else
                if DEBUG
                    println("Attempt $count: Checking box $check and found $(a[check])")    
                end       
                check = a[check]               
            end
        end
           
        if count > ATTEMPTS
            if DEBUG
                println("Prisoner $i failed to find his number in 50 attempts")
            end
            return false
        end
    end
    return true
end

function main()
    tries = 10000000
    success = 0.0
    for i in 1:tries
        if tryit()
            success += 1
        end
    end
    println("Ratio of success = $(success / tries)")
end

main()
 

File: prisoners/prisoners.py
import numpy as np

DEBUG = False
MAX_ATTEMPTS = 50
PRISONERS = 100
 
def tryit():
    a = np.arange(PRISONERS)
    np.random.shuffle(a)
    for i in range(PRISONERS):
        if DEBUG:
            print(f'prisoner {i}')
        count = 0
        check = i
        while count <= MAX_ATTEMPTS:
            count += 1
            if a[check] == i:
                if DEBUG:
                    print(f'Attempt {count}: Checking box {check} and found my number')
                break
            else:
                if DEBUG:
                    print(f'Attempt {count}: Checking box {check} and found {a[check]}')           
                check = a[check]
           
        if count > MAX_ATTEMPTS:
            if DEBUG:
                print(f'Prisoner {i} failed to find his number in {MAX_ATTEMPTS} attempts')
            return False
       
    return True
 
tries = 100000
success = 0
for i in range(tries):
    if tryit():
        success += 1
       
print(f'Ratio of success = {success / tries}')